### DISCUSSION TITLE
References in the Loaded type differ when fields are specified

### DISCUSSION THREAD

**Discussion by S3bt3r** [mikro-orm/mikro-orm#5292] [General]
### Describe the bug

If fields are specified in a find, the return type Loaded works differently. This means that the return cannot be processed uniformly in subsequent functions such as a mapping. Before version 6, this did not seem to be a problem, similar to https://github.com/mikro-orm/mikro-orm/issues/5261

### Reproduction

```ts
async getPostingWithFields(id: number): Promise<PostingDto | null> {
  const posting = await this.#postingRepo.findOne(
    { id },
    {
      fields: ['id', 'thread', 'writtenBy'],
    },
  )

  if (!posting) {
    return null
  }

  return {
    id: posting.id,
    threadId: posting.thread.$.id,
    writtenById: posting.writtenBy.$.id,
    // Does not work for this case
    // threadId: posting.thread.id,
    // writtenById: posting.writtenBy.id,
  }

  // return mapToDto(posting)
}

async getPostingWithoutFields(id: number): Promise<PostingDto | null> {
  const posting = await this.#postingRepo.findOne({ id })

  if (!posting) {
    return null
  }

  return {
    id: posting.id,
    threadId: posting.thread.id,
    writtenById: posting.writtenBy.id,
    // Does not work for this case
    // threadId: posting.thread.$.id,
    // writtenById: posting.writtenBy.$.id,
  }

  // return mapToDto(posting)
}

mapToDto(
  posting:
    | Loaded<Posting, never, 'id' | 'thread' | 'writtenBy', never>
    | Loaded<Posting, never, '*', never>,
): PostingDto {
  return {
    id: posting.id,
    // Does not work for both cases
    // threadId: posting.thread.id,
    // writtenById: posting.writtenBy.id,
    // threadId: posting.thread.$.id,
    // writtenById: posting.writtenBy.$.id,
  }
}
```

### What driver are you using?

@mikro-orm/mysql

### MikroORM version

6.1.5

### Node.js version

18

### Operating system

Debian

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
This is on purpose, [partial loading is now strict](https://mikro-orm.io/docs/upgrading-v5-to-v6#strict-partial-loading), you can't work with such results as if they were full - they are not, the properties you don't select don't exist on the type at all.

I am not entirely sure what you are reporting here, but it seems like you struggle with how TS works in general, not that it would be a bug in the ORM. If you provide a complete repro I can try it and help a bit more, but it needs to be complete, not a snippet like this.

One issue I can see in your snippet is using the union type in the last part with `mapToDto` method definition - that means you need to conditionally handle what type it is, not that you will have access to both types at the same time - that would need to be a type intersection, not union. But maybe its not a problem as `Loaded<Posting, never, '*', never>` should cover the other more specific type already, so the union is in fact just the `Loaded<Posting, never, '*', never>` type.

**Reply by B4nan** [reply]
One more note, the fact that the `$` symbol works in all those cases is a type level bug, caused by the inference of `populate` hint from `fields` hint. You can get around that by specifying empty populate hint. I hope that one could be fixed, as its relevant for to-one relations and we can distinguish those on type level already.

**Reply by S3bt3r** [reply]
I think you are overlooking the fact that in the example the union includes the required properties in both cases. That means in the unspecified (`*`) as well as in the specified (`'id' | 'thread' | 'writtenBy'`) case `thread` is present. Therefore, it should make no difference for the further program flow.

However, this is the case because different types are used for `thread` depending on the case. In the specified case it is `LoadedReference<Loaded<Thread, never, never, never>>` and in the unspecified case  `{ id: number } & Reference<Thread & object>`. This means that in the unspecified case I cannot use `$` but the `id` and in the specified case it is the other way around.

Fun fact: Both types lie. The debugger shows that the `$` and the `id` are present in both cases.

![Screenshot 2024-02-28 121034](https://github.com/mikro-orm/mikro-orm/assets/7798041/175edb9c-5ee5-43fb-b439-a9d14685fee7)
![Screenshot 2024-02-28 121352](https://github.com/mikro-orm/mikro-orm/assets/7798041/00328f58-f46b-4c9b-8dea-a14ec9d772b2)

**Reply by S3bt3r** [reply]
As an addition, in both cases I only want to access the `id` of `thread` and not the complete object, so `populate` should not play a role.

**Reply by B4nan** [reply]
Well, as I said, without a complete repro I can actually try, its really hard to see anything from this.

> Fun fact: Both types lie. The debugger shows that the $ and the id are present in both cases.

Yes, it's there, always, that's on purpose, to protect you from potential runtime errors. This whole feature is about type level, its runtime implementation is not relevant. The `$` property is always the same thing as the class you get it from. And a PK is always present on a reference, even when its not loaded.
